snuba/utils/streams/abstract.py

```diff
@@ -0,0 +1,137 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Callable, Generic, Mapping, Optional, Sequence, TypeVar
+
+
+# The ``TStream`` represents a stream of messages. This should not be a class
+# used for interacting with the stream (that type of interaction should happen
+# via the ``Consumer``), but this type should provide a means for the consumer
+# to uniquely identify a stream. For example, if the stream backend is a file,
+# this type should be the filename -- not the file object or file descriptor.
+# If the stream is a Kafka partition, this should be the topic name and
+# partition ID.
+TStream = TypeVar("TStream")
+
+# The ``TOffset`` represents a single message's position in the stream. For a
+# file, this would be the seek position, or for a Kafka partition, this would
+# be the partition offset. Offset types should have a total ordering.
```

Comment:
I guess a sort of silly example would be the LSN positions from Postgres which have a total ordering (and I think are numbers) but not represented as such to the end user. "Offset types should have a total ordering" might be a bit misleading because they should have a total ordering to the *backend*, but nothing we do here expects them to be directly comparable.

I don't really think it adds that much complexity (and like that it forces this to be treated as opaque in many cases, as it should be) but am not extremely tied to that design.

Reply:
> it forces this to be treated as opaque in many cases, as it should be

A good example of this is this chunk of Kafka-specific code (which will be moved soon): https://github.com/getsentry/snuba/blob/dd9d2dce85c2de40e4bd6858bb7afdbf558d2450/snuba/utils/streams/batching.py#L285-L299

That shouldn't be allowed to be implemented in the batching consumer in the future when this line replaces `KafkaConsumer` with `Consumer`: https://github.com/getsentry/snuba/blob/dd9d2dce85c2de40e4bd6858bb7afdbf558d2450/snuba/utils/streams/batching.py#L106 

At that point it'd be comparing an unbound `TOffset` type variable with an integer, which should cause the type checker to throw a fit.

I do agree that it's not an absolutely necessary generalization, but I do think it's a useful one.

Reply:
A more realistic example would be if we have a `FileConsumer` for testing that reads fixture data from a file, infinitely looping through the file (e.g. for performance testing.) In this case, `TOffset` would actually be a tuple, where the first member would be the loop counter, and the second member would be the seek position in the file.